### PR TITLE
Allows using * to configure the welcome message for all teams 

### PR DESCRIPTION
**Pull request by maisnamrajusingh**
#### Summary
Allows using * to configure the welcome message for all teams 
#### Ticket Link
https://github.com/mattermost/mattermost-plugin-welcomebot/issues/36

### CODE REVIEW

**Review by hanzei — COMMENTED** [review]
The implementation is fine for `"*"`.

The original ticket discussed adding support for Globbing in general and allow patterns like `qa-team-*`. Can this be implemented in a reasonable amount of time and without adding much complexity?

**Comment by maisnamrajusingh** [comment]
@hanzei I think it adds more complexity to the code base, there is already another pr that would let users use "team-1,team2" etc to allow the message for multiple teams. But let me know if you feel I should look into it

File: server/hooks.go [hunk 1]
```diff
@@ -20,6 +20,8 @@ func (p *Plugin) UserHasJoinedTeam(c *plugin.Context, teamMember *model.TeamMemb
 	for _, message := range p.getWelcomeMessages() {
 		if message.TeamName == data.Team.Name {
 			go p.processWelcomeMessage(*data, *message)
+		} else if message.TeamName == "*" {
+			go p.processWelcomeMessage(*data, *message)
```
Comment (hanzei) [inline server/hooks.go:24]: Nit pick: Replace with switch statement

File: server/welcomebot.go [hunk 2]
```diff
@@ -66,7 +66,7 @@ func (p *Plugin) newSampleMessageTemplate(teamName, userID string) (*MessageTemp
 		return nil, fmt.Errorf("failed to query user %s: %w", userID, err)
 	}
 
-	if data.Team, err = p.API.GetTeamByName(teamName); err != nil {
+	if data.Team, err = p.API.GetTeamByName(strings.ToLower(teamName)); err != nil {
```
Comment (hanzei) [inline server/welcomebot.go:69]: Unrelated change?

**Review by hanzei — COMMENTED** [review]
How would the `/welcomebot preview` command work with this feature?

**Comment by hanzei** [comment]
@aaronrothschild I would love to hear your thoughts on the Globbing discussion above.

**Comment by aaronrothschild** [comment]
From the original ticket "however, i would like to be able to send people a welcome-message after their first login - regardless of the team they join."

To me, there are multiple things at play here: 
- Adding the config for multiple teams vs "all" teams is two different settings IMO.   ie: []allusers?  or [x,y,z] teams? 
- The root issue: Send a user message WHEN THEY JOIN THE SERVER - regardless of team membership really. This is "welcomebot" so, I think it makes sense to have a different/setting message for "All New Users, on any team".  The admin doesn't really want the user to see the message every time they join a team (I think) - they just want to make sure a new user sees a message when they join the server, regardless of which team.  Imagine joining 5 teams and seeing the same welcome message each time - because we added this "globbing" setting, would happen a bunch more IMO. (correct me if the behavior is different than I'm describing)
- When someone, at some point adds another team tot he server....does the "*" mean it affects ALL teams or "ALL teams in place when the welcomebot was configured last"? ie: Will an admin need to remember to the Team to the setting in welcomebot so a user will see the welcome message if they join that new team?  That presents a maintenance issue.

File: server/welcomebot.go (hunk 2 above)
Comment (maisnamrajusingh) [reply server/welcomebot.go]: yup, removed it now

File: server/hooks.go (hunk 1 above)
Comment (maisnamrajusingh) [reply server/hooks.go]: done

**Comment by jasonblais** [comment]
@hanzei on above

**Comment by hanzei** [comment]
@maisnamrajusingh https://github.com/mattermost/mattermost-plugin-welcomebot/pull/80#pullrequestreview-697158090 still seems unresolved.

**Comment by hanzei** [comment]
Thanks for your thoughts @aaronrothschild. Based on that feedback the PR needs to get updated so the welcome message triggers at user creation not team joining.

**Comment by maisnamraju** [comment]
@hanzei  started on this, will update you in the next day or so.

**Review by hanzei — COMMENTED** [review]
I doesn't seem like like https://github.com/mattermost/mattermost-plugin-welcomebot/pull/80#issuecomment-872397857 gets addressed in this PR, does it?

**Comment by maisnamrajusingh** [comment]
> I doesn't seem like like [#80 (comment)](https://github.com/mattermost/mattermost-plugin-welcomebot/pull/80#issuecomment-872397857) gets addressed in this PR, does it?

using the * would produce a maintainence issue as mentioned in the comment and confirmed here 
https://github.com/mattermost/mattermost-plugin-welcomebot/pull/80#issuecomment-906348032

**Comment by hanzei** [comment]
@maisnamrajusingh I don't understand what are you referring to. @mickmister talked in https://github.com/mattermost/mattermost-plugin-welcomebot/pull/80#issuecomment-906348032 about the possible conversion from a plugin to an app. IMO this doesn't impact this ticket at all, does it? The suggestion you made in https://github.com/mattermost/mattermost-plugin-welcomebot/pull/80#issuecomment-906348032 still seems reasonable to me and I would like to see that feature.

**Comment by maisnamraju** [comment]
> @maisnamrajusingh I don't understand what are you referring to. @mickmister talked in [#80 (comment)](https://github.com/mattermost/mattermost-plugin-welcomebot/pull/80#issuecomment-906348032) about the possible conversion from a plugin to an app. IMO this doesn't impact this ticket at all, does it? The suggestion you made in [#80 (comment)](https://github.com/mattermost/mattermost-plugin-welcomebot/pull/80#issuecomment-906348032) still seems reasonable to me and I would like to see that feature.

https://github.com/mattermost/mattermost-plugin-welcomebot/pull/80#issuecomment-872397857 I meant this.

**Comment by ayusht2810** [comment]
@mickmister @hanzei @aaronrothschild We looked into the above comments and thought of an approach.
We will be adding a new field in the config `GlobalWelcomeMessage` to support the welcome message when a user has joined a server. Currently, we don't have any hooks present in the mattermost, which triggers when a user joins a sever, so we will be using the [UserHasBeenCreated](https://developers.mattermost.com/integrate/reference/server/server-reference/#Hooks.UserHasBeenCreated) hook to post the message as soon as the user is created in the welcome bot DM.
Also, there is an edge case here: if an admin adds a config with both `GlobalWelcomeMessage` and `TeamName` (with the mattermost teamname present), we will give preference to posting the per team message in the welcome bot DM.
What are your thoughts on this approach? Let me know if I am missing something here.

**Comment by mickmister** [comment]
> Also, there is an edge case here: if an admin adds a config with both `GlobalWelcomeMessage` and `TeamName` (with the mattermost teamname present), we will give preference to posting the per team message in the welcome bot DM. What are your thoughts on this approach? Let me know if I am missing something here.

@ayusht2810 I'm thinking we should create posts for both the global and per-team in this case

**Comment by ayusht2810** [comment]
Created PR https://github.com/mattermost/mattermost-plugin-welcomebot/pull/131 according to the above comments

**Comment by raghavaggarwal2308** [comment]
Closing because of #131
